File: BlClient/app/src/main/java/com/mshare/blclient/MainActivity.java
package com.mshare.blclient;

import androidx.appcompat.app.AppCompatActivity;

import android.Manifest;
import android.annotation.SuppressLint;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothManager;
import android.bluetooth.BluetoothProfile;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanFilter;
import android.bluetooth.le.ScanResult;
import android.bluetooth.le.ScanSettings;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.ParcelUuid;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import java.nio.charset.StandardCharsets;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "ble_client";
    private BluetoothAdapter mBluetoothAdapter;
    private Boolean mScanning = false;
    private Boolean mConnected = false;
    private Boolean mInitialized = false;
    //private HashMap<String, BluetoothDevice> mScanResults;
    private BtleScanCallback mScanCallback;
    private BluetoothLeScanner mBluetoothLeScanner;
    private BluetoothGatt mGatt;
    private Handler mHandler;
    private static final int REQUEST_ENABLE_BT = 0x123;
    private static final int REQUEST_FINE_LOCATION = 0x124;
    private static final int SCAN_PERIOD = 5000;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(BLUETOOTH_SERVICE);
        assert bluetoothManager != null;
        mBluetoothAdapter = bluetoothManager.getAdapter();
        findViewById(R.id.send).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                sendMessage();
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (!getPackageManager().hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)) {
            finish();
        }
        startScan();
    }

    @Override
    protected void onPause() {
        super.onPause();
        disconnectGattServer();
    }

    private void startScan() {
        if (!hasPermissions() || mScanning) {
            return;
        }
        Log.d(TAG, "startScan");
        List<ScanFilter> filters = new ArrayList<>();
        ScanFilter scanFilter = new ScanFilter.Builder()
                .setServiceUuid(new ParcelUuid(Constants.SERVICE_UUID))
                .build();
        filters.add(scanFilter);
        ScanSettings settings = new ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_POWER)
                .build();
        /*mScanResults = new HashMap<>();
        mScanCallback = new BtleScanCallback(mScanResults);*/
        mScanCallback = new BtleScanCallback();
        mBluetoothLeScanner = mBluetoothAdapter.getBluetoothLeScanner();
        mBluetoothLeScanner.startScan(filters, settings, mScanCallback);
        mScanning = true;
        mHandler = new Handler();
        mHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                stopScan();
            }
        }, SCAN_PERIOD);

    }
    private boolean hasPermissions() {
        if (mBluetoothAdapter == null || !mBluetoothAdapter.isEnabled()) {
            requestBluetoothEnable();
            return false;
        } else if (!hasLocationPermissions()) {
            requestLocationPermission();
            return false;
        }
        return true;
    }
    private void requestBluetoothEnable() {
        Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
        startActivityForResult(enableBtIntent, REQUEST_ENABLE_BT);
        Log.d(TAG, "Requested user enables Bluetooth. Try starting the scan again.");
    }
    private boolean hasLocationPermissions() {
        return checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;
    }
    private void requestLocationPermission() {
        requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, REQUEST_FINE_LOCATION);
    }

    private class BtleScanCallback extends ScanCallback {
        /*private HashMap<String, BluetoothDevice> mScanResults;
        private BtleScanCallback(HashMap scanResults) {
            mScanResults = scanResults;
            mScanResults.clear();
        }*/
        @Override
        public void onScanResult(int callbackType, ScanResult result) {
            addScanResult(result);
        }
        @Override
        public void onBatchScanResults(List<ScanResult> results) {
            for (ScanResult result : results) {
                addScanResult(result);
            }
        }
        @Override
        public void onScanFailed(int errorCode) {
            Log.e(TAG, "BLE Scan Failed with code " + errorCode);
        }
        private void addScanResult(ScanResult result) {
            Log.d(TAG, "addScanResult:" + result);
            /*BluetoothDevice device = result.getDevice();
            String deviceAddress = device.getAddress();
            mScanResults.put(deviceAddress, device);*/
            stopScan();
            BluetoothDevice bluetoothDevice = result.getDevice();
            connectDevice(bluetoothDevice);
        }
    }

    private void stopScan() {
        if (mScanning && mBluetoothAdapter != null && mBluetoothAdapter.isEnabled() && mBluetoothLeScanner != null) {
            mBluetoothLeScanner.stopScan(mScanCallback);
            scanComplete();
        }

        mScanCallback = null;
        mScanning = false;
        mHandler = null;
    }

    private void scanComplete() {
        /*if (mScanResults.isEmpty()) {
            return;
        }
        for (String deviceAddress : mScanResults.keySet()) {
            Log.d(TAG, "Found device: " + deviceAddress);
        }*/
    }

    private class GattClientCallback extends BluetoothGattCallback {
        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            super.onConnectionStateChange(gatt, status, newState);
            if (status == BluetoothGatt.GATT_FAILURE) {
                disconnectGattServer();
                return;
            } else if (status != BluetoothGatt.GATT_SUCCESS) {
                disconnectGattServer();
                return;
            }
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                mConnected = true;
                gatt.discoverServices();
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                disconnectGattServer();
            }
        }

        @Override
        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
            super.onServicesDiscovered(gatt, status);
            Log.d(TAG, "onServicesDiscovered status:" + status);
            if (status != BluetoothGatt.GATT_SUCCESS) {
                return;
            }
            BluetoothGattService service = gatt.getService(Constants.SERVICE_UUID);
            BluetoothGattCharacteristic characteristic = service.getCharacteristic(Constants.CHARACTERISTIC_UUID);
            characteristic.setWriteType(BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT);
            mInitialized = gatt.setCharacteristicNotification(characteristic, true);
        }

        @Override
        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            super.onCharacteristicChanged(gatt, characteristic);
            byte[] messageBytes = characteristic.getValue();
            /*for(int i = 0, j = messageBytes.length -1; i < j; ++i, --j) {
                byte temp = messageBytes[i];
                messageBytes[i] = messageBytes[j];
                messageBytes[j] = temp;
            }*/
            String messageString = new String(messageBytes, StandardCharsets.UTF_8);
            Log.d(TAG,"Received message: " + messageString);
            setReceivedData(messageString);
        }
    }

    private void connectDevice(BluetoothDevice device) {
        GattClientCallback gattClientCallback = new GattClientCallback();
        mGatt = device.connectGatt(this, false, gattClientCallback);
    }

    public void disconnectGattServer() {
        mConnected = false;
        if (mGatt != null) {
            mGatt.disconnect();
            mGatt.close();
        }
    }

    private void sendMessage() {
        if (!mConnected || !mInitialized) {
            return;
        }
        EditText messageEditText = findViewById(R.id.data);
        BluetoothGattService service = mGatt.getService(Constants.SERVICE_UUID);
        BluetoothGattCharacteristic characteristic = service.getCharacteristic(Constants.CHARACTERISTIC_UUID);
        String message = messageEditText.getText().toString();
        if(message.isEmpty()) {
            return;
        }
        messageEditText.setText("");
        byte[] messageBytes = message.getBytes(StandardCharsets.UTF_8);
        characteristic.setValue(messageBytes);
        boolean success = mGatt.writeCharacteristic(characteristic);
        Log.d(TAG, "send message success:" + success);
    }

    private String mReceiveData;
    private void setReceivedData(final String data) {
        String time = SimpleDateFormat.getDateTimeInstance().format(new Date(System.currentTimeMillis()));
        if(mReceiveData == null) {
            mReceiveData = time + " " + data;
        } else {
            mReceiveData = time + " " + data + "\n" + mReceiveData;
        }
        final TextView receivedTxt = findViewById(R.id.received);
        runOnUiThread(new Runnable() {
            @SuppressLint("SetTextI18n")
            @Override
            public void run() {
                receivedTxt.setText("Notified data:\n" + mReceiveData);
            }
        });
    }
}

File: BlServer/app/src/main/java/com/mshare/server/MainActivity.kt
package com.mshare.server

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.AdvertiseCallback
import android.bluetooth.le.AdvertiseData
import android.bluetooth.le.AdvertiseSettings
import android.bluetooth.le.BluetoothLeAdvertiser
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.ConnectivityManager
import android.net.wifi.WifiManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.ParcelUuid
import android.util.Log
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.zxing.BarcodeFormat
import com.google.zxing.EncodeHintType
import com.google.zxing.MultiFormatWriter
import com.google.zxing.WriterException
import java.lang.reflect.Field
import java.lang.reflect.InvocationTargetException
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList


class MainActivity : AppCompatActivity() {
    private lateinit var qrImg:ImageView
    private lateinit var mBleThread:BleServerThread
    private lateinit var wifiThread:WlanClientThread
    private var mBleManager: BluetoothManager? = null
    private var mWifiManager:WifiManager? = null
    private var connectivityManager:ConnectivityManager? = null
    private var mBleAdapter: BluetoothAdapter? = null
    private lateinit var mBluetoothLeAdvertiser:BluetoothLeAdvertiser
    private lateinit var mAdvertiseCallback: AdvertiseCallback
    private val mDevices = ArrayList<BluetoothDevice>()

    private val mDataHandler = Handler {
        val data:String = it.obj as String
        wifiThread = WlanClientThread(data,mWifiManager as WifiManager, connectivityManager as ConnectivityManager)
        wifiThread.start()
        true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        qrImg = findViewById(R.id.qrcode)
        //runOnUiThread(Runnable {
            createQrImg()
        //})
        //MyUtils.setImage(qrImg, Constants.BASE_MAC)
        initBleManager()
        mBleThread = BleServerThread(mBleAdapter as BluetoothAdapter, mDataHandler)
        if(Constants.BLE_CONNECT_MODE == 2) {
            mBleThread.start()
        }
        if(mWifiManager == null) {
            mWifiManager = applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
        }
        connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        //wifiThread = WlanClientThread("",mWifiManager as WifiManager)
        //wifiThread.start()
    }
    private fun initBleManager() {
        if (mBleManager == null) {
            mBleManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
            if (mBleAdapter == null && mBleManager != null) {
                mBleAdapter = (mBleManager as BluetoothManager).adapter
                Log.d("client", "mBleAdapter:$mBleAdapter")
            }
        }
    }

    private fun createQrImg() {
        try {
            var mac = getBluetoothMacAddress()
            Log.d("wenpd", "mac:$mac")
            if(mac.isNullOrEmpty() || Constants.FOR_BORAD_TEST) {
                mac = Constants.BASE_BORAD_MAC
            }
            val hints = mutableMapOf<EncodeHintType, String>()
            hints[EncodeHintType.CHARACTER_SET] =  "UTF-8"
            val result = MultiFormatWriter().encode(mac, BarcodeFormat.QR_CODE,350,350,hints)//通过字符串创建二维矩阵
            val width = result.width
            val height = result.height
            Log.d("client", "width:$width height:$height")
            val pixels = Array(width * height){0}

            for (y in 0 until height) {
                val offset = y * width
                for (x in 0 until width) {
                    pixels[offset + x] = if(result.get(x, y)) MyUtils.BLACK else MyUtils.WHITE
                }
            }

            val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)//创建位图
            bitmap.setPixels(pixels.toIntArray(), 0, width, 0, 0, width, height)//设置位图像素集为数组
            qrImg.setImageBitmap(bitmap)//显示二维码
        } catch (e: WriterException) {
            e.printStackTrace()
        }
    }

    @SuppressLint("HardwareIds")
    private fun getBluetoothMacAddress(): String? {
        val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        var bluetoothMacAddress: String? = ""
        if(bluetoothAdapter == null) {
            return null
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            try {
                val mServiceField: Field = bluetoothAdapter.javaClass.getDeclaredField("mService")
                mServiceField.isAccessible = true
                val btManagerService: Any = mServiceField.get(bluetoothAdapter)?:return null
                bluetoothMacAddress =
                        btManagerService.javaClass.getMethod("getAddress").invoke(btManagerService) as String
            } catch (e: NoSuchFieldException) {
            } catch (e: NoSuchMethodException) {
            } catch (e: IllegalAccessException) {
            } catch (e: InvocationTargetException) {
            }
        } else {
            bluetoothMacAddress = bluetoothAdapter.address
        }
        return bluetoothMacAddress
    }

    override fun onResume() {
        super.onResume()
        if (mBleAdapter == null || !(mBleAdapter as BluetoothAdapter).isEnabled) {
            val enableBtIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
            startActivity(enableBtIntent)
            finish()
            return
        }
        if (!packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)) {
            finish()
            return
        }
        if(Constants.BLE_CONNECT_MODE == 1) {
            if (!(mBleAdapter as BluetoothAdapter).isMultipleAdvertisementSupported) {
                finish()
                return
            }
            mBluetoothLeAdvertiser = (mBleAdapter as BluetoothAdapter).bluetoothLeAdvertiser
            mGattServer = mBleManager?.openGattServer(this, gattServerCallback) as BluetoothGattServer
            setupServer()
            startAdvertising()
        }
    }

    override fun onPause() {
        super.onPause()
        if(Constants.BLE_CONNECT_MODE == 1) {
            stopAdvertising()
            stopServer()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        mBleThread.close()
    }

    private fun startAdvertising() {
        val settings = AdvertiseSettings.Builder()
            .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_BALANCED)
            .setTimeout(0)
            .setConnectable(true)
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_LOW)
            .build()
        val advertiseData = AdvertiseData.Builder()
            .setIncludeDeviceName(true)
            .addServiceUuid(ParcelUuid(Constants.BLE_SERVICE_UUID))
            .build()
        /*val scanResponseData = AdvertiseData.Builder()
            .addServiceUuid(ParcelUuid(Constants.BLE_SERVICE_UUID))
            .setIncludeTxPowerLevel(true)
            .build()*/
        mAdvertiseCallback = object : AdvertiseCallback() {
            override fun onStartSuccess(settingsInEffect: AdvertiseSettings) {
                Log.d("wenpd", "BLE advertisement added successfully")
            }

            override fun onStartFailure(errorCode: Int) {
                Log.e("wenpd", "Failed to add BLE advertisement, reason: $errorCode")
            }
        }
        mBluetoothLeAdvertiser.startAdvertising(settings, advertiseData, mAdvertiseCallback)
    }

    private lateinit var mGattServer:BluetoothGattServer
    private fun setupServer() {
        val gattService = BluetoothGattService(Constants.BLE_SERVICE_UUID, BluetoothGattService.SERVICE_TYPE_PRIMARY)
        val characteristicRead = BluetoothGattCharacteristic(Constants.BLE_READ_UUID, BluetoothGattCharacteristic.PROPERTY_READ, BluetoothGattCharacteristic.PERMISSION_READ)
        val descriptor = BluetoothGattDescriptor(Constants.BLE_DESC_UUID, BluetoothGattCharacteristic.PERMISSION_WRITE)
        characteristicRead.addDescriptor(descriptor)
        gattService.addCharacteristic(characteristicRead)
        val characteristicWrite = BluetoothGattCharacteristic(Constants.BLE_WRITE_UUID, BluetoothGattCharacteristic.PROPERTY_WRITE or
                BluetoothGattCharacteristic.PROPERTY_READ or BluetoothGattCharacteristic.PROPERTY_NOTIFY,
                BluetoothGattCharacteristic.PERMISSION_WRITE)
        gattService.addCharacteristic(characteristicWrite)
        Log.d("wenpd", "startGattServer:stagattServicetus=$gattService")
        mGattServer.addService(gattService)
    }

    private val gattServerCallback = object:BluetoothGattServerCallback() {
        override fun onConnectionStateChange(device: BluetoothDevice?, status: Int, newState: Int) {
            super.onConnectionStateChange(device, status, newState)
            Log.d("wenpd", "onConnection:status=$status,newState=$newState")
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                mDevices.add(device!!)
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                mDevices.remove(device)
            }
        }

        override fun onServiceAdded(status: Int, service: BluetoothGattService?) {
            super.onServiceAdded(status, service)
            Log.d("wenpd", "onServiceAdded:status=$status service=$service")
        }

        override fun onCharacteristicReadRequest(device: BluetoothDevice?, requestId: Int, offset: Int, characteristic: BluetoothGattCharacteristic?) {
            super.onCharacteristicReadRequest(device, requestId, offset, characteristic)
            Log.d("wenpd", "onCharacteristicRead:characteristic=${characteristic?.uuid}")
        }

        override fun onCharacteristicWriteRequest(device: BluetoothDevice?, requestId: Int, characteristic: BluetoothGattCharacteristic?, preparedWrite: Boolean, responseNeeded: Boolean, offset: Int, value: ByteArray?) {
            super.onCharacteristicWriteRequest(device, requestId, characteristic, preparedWrite, responseNeeded, offset, value)
            Log.d("wenpd", "onCharacteristicWrite:characteristic=${characteristic?.uuid},value=$value")
            if (characteristic!!.uuid == Constants.BLE_WRITE_UUID) {
                setReceivedData(String(value!!, StandardCharsets.UTF_8));
                mGattServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, 0, null)
                val length = value!!.size
                val reversed = ByteArray(length)
                for (i in 0 until length) {
                    reversed[i] = value[length - (i + 1)]
                }
                characteristic.value = reversed
                for (dev in mDevices) {
                    mGattServer.notifyCharacteristicChanged(dev, characteristic, false)
                }
            }
        }

        override fun onDescriptorWriteRequest(device: BluetoothDevice?, requestId: Int, descriptor: BluetoothGattDescriptor?, preparedWrite: Boolean, responseNeeded: Boolean, offset: Int, value: ByteArray?) {
            super.onDescriptorWriteRequest(device, requestId, descriptor, preparedWrite, responseNeeded, offset, value)
            mGattServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, value)
            Log.d("wenpd", "onDescriptorWrite:value=$value")
        }
    }

    private fun stopServer() {
        mGattServer.close()
    }

    private fun stopAdvertising() {
        mBluetoothLeAdvertiser.stopAdvertising(mAdvertiseCallback)
    }

    private var mReceiveData: String? = null
    private fun setReceivedData(data: String) {
        val time = SimpleDateFormat.getDateTimeInstance()
            .format(Date(System.currentTimeMillis()))
        mReceiveData = if (mReceiveData == null) {
            "$time $data"
        } else {
            "$time $data\n$mReceiveData"
        }
        val receivedTxt = findViewById<TextView>(R.id.received)
        runOnUiThread { receivedTxt.text = "Received data:\n$mReceiveData" }
    }
}
